src/modules/sih/sih.cpp

```diff
@@ -77,15 +85,90 @@ Sih::Sih() :
 	if (_sys_ctrl_alloc.get()) {
 		_actuator_out_sub = uORB::Subscription{ORB_ID(actuator_outputs_sim)};
 	}
+
+#if defined(ENABLE_LOCKSTEP_SCHEDULER)
+	lockstep_loop();
+#else
+	realtime_loop();
+#endif
+	exit_and_cleanup();
 }
 
-Sih::~Sih()
+#if defined(ENABLE_LOCKSTEP_SCHEDULER)
+// Get current timestamp in microseconds
+uint64_t micros()
 {
-	perf_free(_loop_perf);
-	perf_free(_loop_interval_perf);
+	struct timeval t;
+	gettimeofday(&t, nullptr);
+	return t.tv_sec * ((uint64_t)1000000) + t.tv_usec;
+}
+
+void Sih::lockstep_loop()
+{
+
+	int rate = math::min(_imu_gyro_ratemax.get(), _imu_integration_rate.get());
+
+	// default to 400Hz (2500 us interval)
+	if (rate <= 0) {
+		rate = 400;
+	}
+
+	// 200 - 2000 Hz
+	int sim_interval_us = math::constrain(int(roundf(1e6f / rate)), 500, 5000);
+
+	float speed_factor = 1.f;
+	const char *speedup = getenv("PX4_SIM_SPEED_FACTOR");
+
+	if (speedup) {
+		speed_factor = atof(speedup);
+	}
+
+	int rt_interval_us = int(roundf(sim_interval_us / speed_factor));
+
+	PX4_INFO("Simulation loop with %d Hz (%d us sim time interval)", rate, sim_interval_us);
+	PX4_INFO("Simulation with %.1fx speedup. Loop with (%d us wall time interval)", (double)speed_factor, rt_interval_us);
+
+	if (_lockstep_component < 0) {
+		_lockstep_component = px4_lockstep_register_component();
```

Comment:
This should be removed as this is the task waiting for others

Reply:
Hmm, I have this from here https://github.com/PX4/PX4-Autopilot/blob/3aee2497bbe63090ec220fb23c85ff9e24871c57/src/modules/simulator/simulator_mavlink.cpp#L485  originally, makes sense to me that the work queue also has to wait for the simulation to finish before continuing. But I agree, when the work queue items already check for new data in their code, so they will be blocked until new data arrives. Seems that this was added for multi-ekf support

Reply:
Thanks @bkueng , the reason for the lockstep component in simulator_mavlink is because it is on another thread which needs to be scheduled in as well. I'll remove it here